### DISCUSSION TITLE
C++ Not Allowed?

### DISCUSSION THREAD

**Discussion by NamjaeChoi** [CEED/libCEED#1684] [Q&A] [answered]
Hello,

According to the document,

``
The entire contents of this file and all locally included files (#include "foo.h") are used during JiT compilation for GPU backends. Installed headers (#include <bar.h>) are omitted in the source code passed to JiT, but the compilation environment may supply common headers such as <math.h>. These source file must only contain syntax constructs supported by C99 and all targeted backends (i.e. CUDA for /gpu/cuda, OpenCL/SYCL for /gpu/sycl, etc.).
``

It sounds like using class syntax is not allowed in QFunctions at all. Is my understanding true? What is limiting us to C99?

Thank you.

**Comment by jeremylt** [comment]
Do you have something specific in mind? I'm not seeing an obvious use case where a plain struct wouldn't work as well as a class, but I also don't use a lot of C++?

Some compilation environments (OCCA, SYCL) are pickier, but if you only care about one environment, you can go with whatever that backend successfully compiles, in practice.

**Comment by NamjaeChoi** [comment]
We are exploring if libCEED can be integrated with MOOSE. MOOSE is object-oriented where the user derives its own classes from base MOOSE classes. We were thinking of passing these objects as contexts to QFunctions, typecast them, and call user methods (the user methods might be precompiled, not JIT compiled). So I guess C99 is not a strict restriction and dependent on the backends?

**Reply by jeremylt** [reply, marked as answer]
Yeah, if you don't care about the OCCA and SYCL backends, it gets a lot easier to experiment. The CPU side just cares about getting a good function pointer for the QFunction and CUDA and HIP can handle whatever syntax CUDA and HIP normally support. The entire definition of that class and its methods would need to be accessible to the JiT backend for compiling, which is where you might run into compile time issues if the entirety of those files with those definitions aren't CUDA/HIP friendly.

**Reply by NamjaeChoi** [reply]
Okey, thanks for the reply. That makes us feel more comfortable.

**Reply by jedbrown** [reply]
OpenCL is really the sticking point.  SYCL does not yet have an "online compiler" so the SYCL backend uses OpenCL (Intel is doing some work toward a SYCL-native online compiler so that may change in the next year or two). If you don't care about SYCL and you would like to use C++ in qfunctions that are included from C++ source, you're welcome to.
